manticore/core/cpu/abstractcpu.py

```diff
@@ -742,36 +742,36 @@ def execute(self):
             return
 
         name = self.canonicalize_instruction_name(insn)
-
-        def fallback_to_emulate(*operands):
-            text_bytes = ' '.join('%02x'%x for x in insn.bytes)
-            logger.info("Unimplemented instruction: 0x%016x:\t%s\t%s\t%s",
-                        insn.address, text_bytes, insn.mnemonic, insn.op_str)
-            self.emulate(insn)
-
-        implementation = getattr(self, name, fallback_to_emulate)
-
+            
         if logger.level == logging.DEBUG :
             logger.debug(self.render_instruction(insn))
             for l in self.render_registers():
                 register_logger.debug(l)
 
-        #FIXME(yan): In the case the instruction implementation invokes a system call, we would not be able to
-        # publish the did_execute_instruction event from here, so we capture and attach it to the syscall
-        # exception for the platform to emit it for us once the syscall has successfully been executed.
-        def did_exec():
-            self._icount += 1
-            self._publish('did_execute_instruction', self._last_pc, self.PC, insn)
-
         try:
-            implementation(*insn.operands)
+            try:
+                getattr(self, name)(*insn.operands)
+            except AttributeError:
+                text_bytes = ' '.join('%02x'%x for x in insn.bytes)
+                logger.info("Unimplemented instruction: 0x%016x:\t%s\t%s\t%s",
+                            insn.address, text_bytes, insn.mnemonic, insn.op_str)
+                self.emulate(insn)
         except (Interruption, Syscall) as e:
-            e.on_handled = did_exec
+            e.on_handled = lambda: self._publish_instruction_as_executed(insn)
```

Comment:
We should consider handling syscalls with an event callback instead of using an exception. So it continues here after the handler execution?

Reply:
That decision is beyond the scope of my change. My change is mostly superficial, but has a minor performance impact.